File: sdks/ts/memory/src/ratelimit/headers.ts
// SPDX-License-Identifier: Apache-2.0

/**
 * Parse rate-limit headers from HTTP responses. Supports the common
 * header names used by OpenAI, Anthropic, and other LLM providers.
 */

import type { HeaderNameOptions, RateLimitHeaders } from './types.js'

/** Default candidate header names for remaining requests/tokens. */
const DEFAULT_REMAINING_HEADERS = [
  'x-ratelimit-remaining',
  'ratelimit-remaining',
  'x-ratelimit-remaining-requests',
  'x-ratelimit-remaining-tokens',
] as const

/** Default candidate header names for the rate limit ceiling. */
const DEFAULT_LIMIT_HEADERS = [
  'x-ratelimit-limit',
  'ratelimit-limit',
  'x-ratelimit-limit-requests',
  'x-ratelimit-limit-tokens',
] as const

/** Default candidate header names for the reset time. */
const DEFAULT_RESET_HEADERS = [
  'x-ratelimit-reset',
  'ratelimit-reset',
  'x-ratelimit-reset-requests',
  'x-ratelimit-reset-tokens',
] as const

/** Default candidate header names for retry-after. */
const DEFAULT_RETRY_AFTER_HEADERS = ['retry-after'] as const

/** Build a RateLimitHeaders object, omitting undefined properties. */
const buildHeaders = (
  remaining: number | undefined,
  limit: number | undefined,
  resetAt: Date | undefined,
  retryAfter: number | undefined,
): RateLimitHeaders => ({
  ...(remaining !== undefined ? { remaining } : {}),
  ...(limit !== undefined ? { limit } : {}),
  ...(resetAt !== undefined ? { resetAt } : {}),
  ...(retryAfter !== undefined ? { retryAfter } : {}),
})

/**
 * Resolve header name options, using defaults for any unspecified groups.
 */
const resolveNames = (custom?: HeaderNameOptions) => ({
  remaining: custom?.remaining ?? DEFAULT_REMAINING_HEADERS,
  limit: custom?.limit ?? DEFAULT_LIMIT_HEADERS,
  reset: custom?.reset ?? DEFAULT_RESET_HEADERS,
  retryAfter: custom?.retryAfter ?? DEFAULT_RETRY_AFTER_HEADERS,
})

/** Generic getter abstraction — unifies Headers (fetch API) and plain records. */
type HeaderGetter = (name: string) => string | undefined

const getterFromHeaders = (h: Headers): HeaderGetter =>
  (name) => h.get(name) ?? undefined

const getterFromRecord = (r: Readonly<Record<string, string | undefined>>): HeaderGetter =>
  (name) => r[name]

/** Parse rate-limit headers from a fetch API Headers object. */
export const parseRateLimitHeaders = (
  headers: Headers,
  headerNames?: HeaderNameOptions,
): RateLimitHeaders => {
  const names = resolveNames(headerNames)
  const get = getterFromHeaders(headers)
  return buildHeaders(
    firstInt(get, names.remaining),
    firstInt(get, names.limit),
    firstDate(get, names.reset),
    parseRetryAfterValue(get, names.retryAfter),
  )
}

/** Parse rate-limit headers from a plain record (e.g. from node http). */
export const parseRateLimitHeaderRecord = (
  headers: Readonly<Record<string, string | undefined>>,
  headerNames?: HeaderNameOptions,
): RateLimitHeaders => {
  const names = resolveNames(headerNames)
  const get = getterFromRecord(headers)
  return buildHeaders(
    firstInt(get, names.remaining),
    firstInt(get, names.limit),
    firstDate(get, names.reset),
    parseRetryAfterValue(get, names.retryAfter),
  )
}

const firstInt = (get: HeaderGetter, candidates: readonly string[]): number | undefined => {
  for (const name of candidates) {
    const v = get(name)?.trim()
    if (v === undefined || v === '') continue
    const n = Number.parseInt(v, 10)
    if (!Number.isNaN(n)) return n
  }
  return undefined
}

const firstDate = (get: HeaderGetter, candidates: readonly string[]): Date | undefined => {
  for (const name of candidates) {
    const v = get(name)?.trim()
    if (v === undefined || v === '') continue
    // Try Unix epoch seconds.
    const epoch = Number.parseInt(v, 10)
    if (!Number.isNaN(epoch) && String(epoch) === v) {
      return new Date(epoch * 1000)
    }
    // Try HTTP date.
    const d = new Date(v)
    if (!Number.isNaN(d.getTime())) return d
  }
  return undefined
}

const parseRetryAfterValue = (get: HeaderGetter, candidates: readonly string[]): number | undefined => {
  for (const name of candidates) {
    const v = get(name)?.trim()
    if (v === undefined || v === '') continue
    const secs = Number.parseInt(v, 10)
    if (!Number.isNaN(secs)) return secs
    // HTTP date.
    const d = new Date(v)
    if (!Number.isNaN(d.getTime())) {
      const delta = (d.getTime() - Date.now()) / 1000
      return delta > 0 ? Math.ceil(delta) : undefined
    }
  }
  return undefined
}

File: go/ratelimit/headers.go
// SPDX-License-Identifier: Apache-2.0

package ratelimit

import (
	"net/http"
	"strconv"
	"strings"
	"time"
)

// HeaderNames configures which HTTP header names to check when parsing
// rate-limit responses. The zero value uses DefaultHeaderNames.
type HeaderNames struct {
	Remaining  []string
	Limit      []string
	ResetAt    []string
	RetryAfter []string
}

// DefaultHeaderNames returns the standard header names checked when
// parsing rate-limit responses, covering OpenAI, Anthropic, and other
// common LLM provider conventions.
func DefaultHeaderNames() HeaderNames {
	return HeaderNames{
		Remaining: []string{
			"X-Ratelimit-Remaining",
			"X-RateLimit-Remaining",
			"Ratelimit-Remaining",
			"X-RateLimit-Remaining-Requests",
			"X-RateLimit-Remaining-Tokens",
		},
		Limit: []string{
			"X-Ratelimit-Limit",
			"X-RateLimit-Limit",
			"Ratelimit-Limit",
			"X-RateLimit-Limit-Requests",
			"X-RateLimit-Limit-Tokens",
		},
		ResetAt: []string{
			"X-Ratelimit-Reset",
			"X-RateLimit-Reset",
			"Ratelimit-Reset",
			"X-RateLimit-Reset-Requests",
			"X-RateLimit-Reset-Tokens",
		},
		RetryAfter: []string{
			"Retry-After",
		},
	}
}

// defaultNames is the package-level default used when no custom names
// are provided.
var defaultNames = DefaultHeaderNames()

// resolve returns the provided names if non-empty, or the defaults.
func (h HeaderNames) resolve() HeaderNames {
	if len(h.Remaining) == 0 {
		h.Remaining = defaultNames.Remaining
	}
	if len(h.Limit) == 0 {
		h.Limit = defaultNames.Limit
	}
	if len(h.ResetAt) == 0 {
		h.ResetAt = defaultNames.ResetAt
	}
	if len(h.RetryAfter) == 0 {
		h.RetryAfter = defaultNames.RetryAfter
	}
	return h
}

// ParseHeaders extracts rate-limit metadata from an HTTP response.
// Missing or unparseable headers are silently ignored (zero values).
func ParseHeaders(resp *http.Response) Headers {
	if resp == nil {
		return Headers{}
	}
	return ParseHeaderMap(resp.Header)
}

// ParseHeaderMap extracts rate-limit metadata from a raw header map
// using the default header names.
func ParseHeaderMap(h http.Header) Headers {
	return ParseHeaderMapWith(h, HeaderNames{})
}

// ParseHeaderMapWith extracts rate-limit metadata from a raw header map
// using custom header names. Zero-value fields in names fall back to
// the defaults.
func ParseHeaderMapWith(h http.Header, names HeaderNames) Headers {
	if h == nil {
		return Headers{}
	}
	n := names.resolve()
	return Headers{
		Remaining:  firstInt(h, n.Remaining),
		Limit:      firstInt(h, n.Limit),
		ResetAt:    firstTime(h, n.ResetAt),
		RetryAfter: firstDuration(h, n.RetryAfter),
	}
}

// firstInt returns the integer value of the first header that parses
// successfully from the candidate list.
func firstInt(h http.Header, candidates []string) int {
	for _, name := range candidates {
		v := h.Get(name)
		if v == "" {
			continue
		}
		n, err := strconv.Atoi(strings.TrimSpace(v))
		if err == nil {
			return n
		}
	}
	return 0
}

// firstTime returns the time value of the first header that parses
// successfully. Supports both Unix epoch seconds and RFC 1123 format.
func firstTime(h http.Header, candidates []string) time.Time {
	for _, name := range candidates {
		v := strings.TrimSpace(h.Get(name))
		if v == "" {
			continue
		}
		// Try Unix epoch seconds first (most common).
		if epoch, err := strconv.ParseInt(v, 10, 64); err == nil {
			return time.Unix(epoch, 0)
		}
		// Try RFC 1123 (used by some CDN providers).
		if t, err := time.Parse(time.RFC1123, v); err == nil {
			return t
		}
	}
	return time.Time{}
}

// firstDuration parses a Retry-After header as either seconds (integer)
// or an HTTP-date (RFC 1123). Returns zero on failure.
func firstDuration(h http.Header, candidates []string) time.Duration {
	for _, name := range candidates {
		v := strings.TrimSpace(h.Get(name))
		if v == "" {
			continue
		}
		// Integer seconds.
		if secs, err := strconv.Atoi(v); err == nil {
			return time.Duration(secs) * time.Second
		}
		// HTTP-date: compute duration from now.
		if t, err := time.Parse(time.RFC1123, v); err == nil {
			d := time.Until(t)
			if d > 0 {
				return d
			}
		}
	}
	return 0
}
